Eval

felipessalvatore opened the pull request:
i) Add decorator to keep time information

ii) I substitute the naive analogy comparison by a vectorized tensorflow implementation. The old "analogy_score" is substituted by the class "Evaluator".  And the class ModelJudge encapsulates the functions "compare_models" and "save_comparison"

iii) Add a method to plot the different metrics in the class ModelJudge using the seaborn library.

One of the plot is the following:

![score](https://user-images.githubusercontent.com/11237428/29755686-e8ae461e-8b71-11e7-89f5-9cb276f37f90.png)

lucasmoura commented on src/eval/Evaluator.py:
```diff
@@ -0,0 +1,160 @@
+                             normalized_emb,
+                             transpose_b=True)
+            _, pred_idx = tf.nn.top_k(dist, self.top_results + 2)
+        with tf.Session(graph=graph) as sess:
```
Maybe the Evaluator class could receive a session object or create one in the __init__ method, by doing so, it would avoid creating a context just to evaluate a variable. I thought about something like:

```python
self.top_k = self.session.run(...)
```

lucasmoura commented on src/eval/Evaluator.py:
```diff
@@ -0,0 +1,160 @@
+        for i, line in enumerate(self.top_k):
+            b_c = list(self.analogy_questions[i][1:3])
+            new = []
+            for idx in line:
```
If analogies b an c are in the top_k array and they have the biggest values, maybe this code can help:

```python
import numpy as np

a = np.arange(10)
np.delete(a, np.argsort(a)[-2:])
```

felipessalvatore replied:
I did not change this part because although in the majority of cases the words b and c have the biggest values, this is not always the case. Using the portuguese corpus we have examples like:

- **analogy** = escrever escrevem trabalhou trabalham
- **top 12 results** = ['escrevem', 'estudavam', 'atuam', 'estrearam', 'regeu', 'bud', 'trabalhou', 'trabalham', 'trabalharam', 'moravam', 'milla', 'atuaram']

Word b ('escrevem') has the biggest value, but word c ('trabalhou') has not the second biggest value.

lucasmoura commented on src/eval/Evaluator.py:
```diff
@@ -0,0 +1,160 @@
+import numpy as np
```
The code has some flake8 warning on it, maybe it is worth to take a look and fix some of it.

If you think some errors are not worth fixing, add a #noqa tag at the end of the line that produced the error, or if it is
a specific warning that you don't want to consider, create .flake8 file and add the command to ignore the warning.

This link provides how you can do that:
[flake8 config](http://flake8.pycqa.org/en/latest/user/configuration.html)

felipessalvatore commented:
Ok, a lot of things have changed.

1) I added a bar plot as required.

2) @ThiagoLira Please look the changes that I did in the Gensim model. I add some tests for this model and try to make it more simpler and more similar to the class TFWord2Vec. @lucasmoura I also changed some minor things about your TFWord2Vec implementation (for example, the get_pickle method was not working). So, I would ask to both of you take a look on the changes that I did.

3) @lucasmoura I saw that the TFWord2Vec do not change the words from a text to the lower case format. I do not like that because it make the vocabulary unnecessary bigger. So I added a function to transform a text in  the lower case form. But it would be nice if this model could do that without this additional function  

4) I raw a little experiment on the doncovin computer:
 
![score](https://user-images.githubusercontent.com/11237428/30004949-5abcd400-90ae-11e7-9592-087dd8e9b44a.png)
 
The figure above is not significant since I trained the tf model with only one epoch (I will use 15 epochs as is indicated in the tensorflow/models repo).

The experiments scripts are set for that computer. We can discuss after how can we generalize these scripts

lucasmoura commented:
I believe we can merge this already. @felipessalvatore if google-compute-engine is really necessary to run travis, I can merge this branch.
